More on collections security (#117)
ietf-rats-wg/draft-ietf-rats-msg-wrap: draft-ietf-rats-msg-wrap.md

## CMW Collections {#cmw-coll}

Layered Attesters and composite devices ({{Sections 3.2 and 3.3 of -rats-arch}}) generate Evidence that consists of multiple parts.
For example, in data center servers, it is not uncommon for separate attesting environments (AE) to serve a subsection of the entire machine.
One AE might measure and attest to what was booted on the main CPU, while another AE might measure and attest to what was booted on a SmartNIC plugged into a PCIe slot, and a third AE might measure and attest to what was booted on the machine's GPU.
To allow aggregation of multiple, potentially non-homogeneous evidence formats collected from different AEs, this document defines a CMW "collection" as a container that holds several CMW items, each with a label that is unique within the scope of the collection.

> Even though initially designed to support the layered Attester and composite device use cases, the CMW collection can be repurposed for other use cases requiring aggregation of RATS conceptual messages.

Reviewer: ```suggestion
Although originally designed to support layered Attester and composite device use cases, the CMW collection can be adapted for other scenarios that require the aggregation of RATS conceptual messages.
```

> For example, collections may be used to carry groups of Endorsements or Reference Values, Attestation Results, and so on.

Reviewer: ```suggestion
For instance, collections may be used to group Endorsements, Reference Values, Attestation Results, and more.
```

> One CMW collection can be a mixture of different types of messages.

Reviewer: ```suggestion
A single CMW collection can contain a mix of different message types, and it can also be used to carry messages related to multiple devices simultaneously.
```

> One CMW collection can be used to carry messages related to more than one device at a time.

Reviewer: ```suggestion
```

{: #fig-cddl-collection artwork-align="left"
   title="CDDL definition of the CMW collection format"}

> CMW itself provides no facilities for authenticity, integrity protection, privacy or attestation security.

Reviewer: ```suggestion
CMW itself provides no facilities for authenticity, integrity protection, or confidentiality.
```

> It is up to the designer of each use case to understand the security properties necessary and add them around the collection.

Reviewer: ```suggestion
It is the responsibility of the designer for each use case to determine the necessary security properties and implement them accordingly.
```

> A secure channel (e.g., using TLS) or some form of object security (e.g., using JWT) may be sufficient, but not always.

Reviewer: ```suggestion
A secure channel (e.g., via TLS) or object-level security (e.g., using JWT) may suffice in some scenarios, but not in all.
```

When a CMW is used to carry the Evidence for composite or layered attestation for a single device, the security properties needed are that of attestation.
In particular, all the members in a CMW must be bound together so that an attacker can not replace one Evidence message showing compromise with that from a non-compromised device.
The authenticity and integrity protection MUST be attestation-oriented.
For further security considerations about collections, see {{seccons-coll}}.

### Relation to EAT `submods`

EAT submods ({{Section 4.2.18 of -rats-eat}}) provide a facility for aggregating attestation that has built-in security and will be suitable for some of the same attestation Evidence use cases covered by CMW collections.
However, compared to CMW collections, EAT submods are limited in two ways:

> 1. Only EAT-formatted payloads are supported,

Author: ```suggestion
1. {{EAT}} allows carrying non-EAT-formatted types by augmenting the $EAT-CBOR-Tagged-Token socket or the $JSON-Selector socket. However, these need to be specified in subsequent standard documents updating the EAT specification,
```

> 2. Their top-down structure does not align well with the bottom-up approach layered attesters use to build the chain of trust, making them not ideal for modelling layered attestation.

Reviewer 2: You can make an EAT submod a format other than EAT by defining a type for it. In CBOR-format EATs, that's just defining a new CBOR tag to identify it. Might be a bit weird to define a CBOR tag to identify an X.509 format attestation, but it is possible.

So I don't think 1. is true.

A more correct statement would be that EAT can't represent layered attestation when the topmost/outermost/lead attestation is not EAT format.

EAT seems fine for most layered attestation (see recent email to rats). Seems to me that CCA is doing rather unusual layered attestation. CCA attestation evidence spans devices, right? That's unusual for attestation IMO and beyond what I imagined anyone doing in EAT (and RATS).

EAT should not be presented as being unsuitable or bad at layered attestation.

Maybe say something like, while EAT submods are fine for most layered attestation, CMW is more flexible and can represent more complex or unusual layered attestation like a bottom-up chain of trust.

Author: > You can make an EAT submod a format other than EAT by defining a type for it. In CBOR-format EATs, that's just defining a new CBOR tag to identify it. Might be a bit weird to define a CBOR tag to identify an X.509 format attestation, but it is possible.
>
> So I don't think 1. is true.

This is not what I understand from reading  [4.2.18. submods (Submodules)](https://ietf-rats-wg.github.io/eat/draft-ietf-rats-eat.html#section-4.2.18).  There is no mention of the possibility of incorporating non-EAT tokens/claims-sets.  

Can you point me to the relevant text?

> A more correct statement would be that EAT can't represent layered attestation when the topmost/outermost/lead attestation is not EAT format.
> 
> EAT seems fine for most layered attestation (see recent email to rats). Seems to me that CCA is doing rather unusual layered attestation. CCA attestation evidence spans devices, right? That's unusual for attestation IMO and beyond what I imagined anyone doing in EAT (and RATS).
> 
> EAT should not be presented as being unsuitable or bad at layered attestation.
> 
> Maybe say something like, while EAT submods are fine for most layered attestation, CMW is more flexible and can represent more complex or unusual layered attestation like a bottom-up chain of trust.

We can certainly curate the text to avoid sounding adversarial vs EAT (that's certainly not my intent, on the contrary, I think EAT and CMW are an excellent pair.)

However, I stand by my assessment that submods are not great at layered attestation because they go counter the "natural" construction of the chain of trust of the device.

Typically (e.g., DICE, Arm CCA), the layer below signs the layer above, not the opposite.

Reviewer 3: Re: incorporating non-EAT tokens/claims-sets, I think Laurence is referring to the fact that $EAT-CBOR-Tagged-Token is extensible. The potential for adding new types is noted in section 4.2.18 as follows:

```
The addition of further types is accomplished by augmenting the $EAT-CBOR-Tagged-Token socket or the $JSON-Selector socket.
```

You are correct the text does not explicitly call out that ASN.1 types, or other, could be accommodated this way but the means to support such is there.

Author: Thanks for the pointer!

Would it be fair to say something like:

> _"{{EAT}} allows the carrying of non-EAT-formatted types by augmenting the `$EAT-CBOR-Tagged-Token` socket or the `$JSON-Selector` socket. However, these need to be specified in subsequent standard documents updating the EAT specification."_

?

Reviewer 3: I think so.

Author: Integrated in [39e1f53](https://github.com/ietf-rats-wg/draft-ietf-rats-msg-wrap/pull/117/commits/39e1f533af004805e30967fba3284101ac4c9a7f)

Author: Taking this a step further, we may as well drop this and add a new subsection _"Transporting CMW in EAT submods"_ that defines the sockets extensions:

```
$EAT-CBOR-Tagged-Token /= #6.TBD(cbor-CMW)
$JSON-Selector /= [type: "CMW", nested-token: json-CMW]
```

Where:

* `cbor-CMW` and `json-CMW` are defined in [Appendix A of CMW](https://www.ietf.org/archive/id/draft-ietf-rats-msg-wrap-08.html#appendix-A)
* `6.TBD` is the CBOR tag [requested for CMW](https://www.ietf.org/archive/id/draft-ietf-rats-msg-wrap-08.html#section-8.3)

@carl-wallace @laurencelundblade @gmandyam would that work?

Reviewer 3: Adding use of submod extensibility here seems like a good idea to me.

> 2. They are not a good match for modelling layered attestation given their top-down structure does not fit the bottom-up way in which the chain of trust is built in layered attestation.

Author: ```suggestion
2. They are not a good fit for modelling layered attestation, as their top-down structure does not align with the bottom-up approach used in layered attestation to build the chain of trust.
```

}
~~~

# Transporting COSE and JOSE Web Tokens

To allow embedding CMWs and CMW collections into CBOR-based protocols and web APIs, this document defines two `"cmw"` claims for JSON Web Tokens (JWT) and CBOR Web Tokens (CWT).

> Their definitions are provided in {{iana-jwt}} and {{iana-cwt}} respectively..

Author: ```suggestion
Their definitions are provided in {{iana-jwt}} and {{iana-cwt}} respectively.
```

## Encoding Requirements

> A CMW collection carried in a `"cmw"` JWT claim ({{iana-jwt}}) MUST be a `json-collection`.
> A CMW collection carried in a `"cmw"` CWT claim ({{iana-cwt}}) MUST be a `cbor-collection`.
> 
> A CMW record carried in a `"cmw"` JWT claim ({{iana-jwt}}) MUST be a `json-record`.
> A CMW record carried in a `"cmw"` CWT claim ({{iana-cwt}}) MUST be a `cbor-record`.

Author: ```suggestion
A CMW collection carried in a `"cmw"` JWT claim MUST be a `json-collection`.
A CMW collection carried in a `"cmw"` CWT claim MUST be a `cbor-collection`.

A CMW record carried in a `"cmw"` JWT claim MUST be a `json-record`.
A CMW record carried in a `"cmw"` CWT claim MUST be a `cbor-record`.
```

}
~~~

> # Transporting COSE and JOSE Web Tokens

Author: ```suggestion
# Transporting CMW in COSE and JOSE Web Tokens
```

> To allow embedding CMWs and CMW collections into CBOR-based protocols and web APIs, this document defines two `"cmw"` claims for JSON Web Tokens (JWT) and CBOR Web Tokens (CWT).

Reviewer: ```suggestion
To facilitate the embedding of CMWs and CMW collections in CBOR-based protocols and web APIs, this document defines two `"cmw"` claims for use with JSON Web Tokens (JWT) and CBOR Web Tokens (CWT).
```

> Their definitions are provided in {{iana-jwt}} and {{iana-cwt}} respectively.

Reviewer: ```suggestion
The definitions for these claims can be found in {{iana-jwt}} and {{iana-cwt}}, respectively.
```

## Encoding Requirements

A CMW collection carried in a `"cmw"` JWT claim MUST be a `json-collection`.
A CMW collection carried in a `"cmw"` CWT claim MUST be a `cbor-collection`.

A CMW record carried in a `"cmw"` JWT claim MUST be a `json-record`.
A CMW record carried in a `"cmw"` CWT claim MUST be a `cbor-record`.

# Transporting CMW in X.509 Messages {#x509}

CMW may need to be transported in PKIX messages, such as Certificate Signing Requests (CSRs) or in X.509 Certificates and Certificate Revocation Lists (CRLs).

> The former use is documented in {{-csr-a}}, the latter in Section 6.1 of {{DICE-arch}}.

Reviewer: ```suggestion
The use of CMW in CSRs is documented in {{-csr-a}}, while its application in X.509 Certificates and CRLs is detailed in Section 6.1 of {{DICE-arch}}.
```

> This section specifies the CMW extension to carry CMW objects.

Reviewer: ```suggestion
This section outlines the CMW extension designed to carry CMW objects.
```

The CMW extension MAY be included in X.509 Certificates, CRLs {{-pkix}}, and CSRs.

The CMW extension MUST be identified by the following object identifier:

~~~asn.1
id-pe-cmw  OBJECT IDENTIFIER ::=
        { iso(1) identified-organization(3) dod(6) internet(1)
          security(5) mechanisms(5) pkix(7) id-pe(1) TBD }
~~~

This extension SHOULD NOT be marked critical.
It MAY be marked critical in cases where the attestation-related information is essential for granting resource access, and there is a risk that legacy relying parties would bypass such controls.

The CMW extension MUST have the following syntax:

~~~asn.1
CMW ::= CHOICE {
    json UTF8String,
    cbor OCTET STRING
}
~~~

> The CMW MUST contain the serialized CMW object in JSON or CBOR format, using the appropriate CHOICE entry.

Reviewer: ```suggestion
The CMW MUST include the serialized CMW object in either JSON or CBOR format, utilizing the appropriate CHOICE entry.
```

The DER-encoded CMW is the value of the OCTET STRING for the extnValue field of the extension.

## ASN.1 Module {#asn1-x509}

This section provides an ASN.1 module {{X.680}} for the CMW extension, following the conventions established in {{RFC5912}} and {{RFC6268}}.

~~~asn.1
CMWExtn
  { iso(1) identified-organization(3) dod(6) internet(1)
    security(5) mechanisms(5) pkix(7) id-mod(0)
    id-mod-cmw-collection-extn(TBD) }

DEFINITIONS IMPLICIT TAGS ::=
BEGIN

IMPORTS
  EXTENSION
  FROM PKIX-CommonTypes-2009  -- RFC 5912
    { iso(1) identified-organization(3) dod(6) internet(1)
      security(5) mechanisms(5) pkix(7) id-mod(0)
      id-mod-pkixCommon-02(57) } ;

-- CMW Extension

ext-CMW EXTENSION ::= {
  SYNTAX CMW
  IDENTIFIED BY id-pe-cmw }

-- CMW Extension OID

id-pe-cmw  OBJECT IDENTIFIER  ::=
   { iso(1) identified-organization(3) dod(6) internet(1)
     security(5) mechanisms(5) pkix(7) id-pe(1) TBD }

-- CMW Extension Syntax

CMW ::= CHOICE {
    json UTF8String,
    cbor OCTET STRING
}

END
~~~

## Compatibility with DICE `ConceptualMessageWrapper`

> Section 6.1.8 of {{DICE-arch}} defines the ConceptualMessageWrapper format and the associated object identifier.

Reviewer: ```suggestion
Section 6.1.8 of {{DICE-arch}} specifies the ConceptualMessageWrapper (CMW) format and its corresponding object identifier.
```

> The CMW format defined in {{DICE-arch}} allows only a subset of the CMW grammar defined in this document.

Reviewer: ```suggestion
The CMW format outlined in {{DICE-arch}} permits only a subset of the CMW grammar defined in this document.
```

> Specifically, the tunnel and collection formats cannot be encoded using DICE CMWs.

Reviewer: ```suggestion
In particular, the tunnel and collection formats cannot be encoded using DICE CMWs.
```

The binding does not necessarily have to be a signature over the CMW collection, it might also be achieved through identifiers, cross-linking, signing or hashing between the members of the collection.
Client-authenticated TLS may be used to bind a CMW collection of Evidence messages.
However, the client key used with TLS should not be that of the end-user or owner of the device.

> Instead, it must be attestation-oriented key material from the device or the attester manufacturer.

Author: ```suggestion
Instead, it should be attestation-oriented key material from the device or the attester manufacturer.
```
